Support archival and restoration of POSIX permissions

MartinLau7 opened the pull request:
#### Submit log
```
1. Unzip to add recovery for macOS file permissions
2. Compress add retention for macOS file permissions
3. UZKFileInfo Add Permission Attributes
```

#### Test Code
```mm
NSString *zipFilePath = @"/Users/martin/
UZKArchive *archive = [[UZKArchive alloc] initWithPath:zipFilePath error:&archiveError];
if (archiveError) {
        return;
}
NSDictionary *infoDict = [[NSFileManager defaultManager] attributesOfItemAtPath:[@"your file" stringByAppendingPathComponent:obj] error:nil];
NSNumber *mode = [infoDict objectForKey:NSFilePosixPermissions];

NSData *fileData = ...;
if (![archive writeData:fileData filePath:@"zip file path" posixPermissions:[mode unsignedLongValue] error:nil]) {
       NSLog(@"zip file error");
}
```

abbeycode commented on Source/UZKArchive.m:
```diff
@@ -280,11 +280,8 @@ - (void)setComment:(NSString *)comment
     }
 }
 
-
```
(Nitpick:) Please remove this whitespace change

abbeycode commented on Source/UZKArchive.m:
```diff
@@ -280,11 +280,8 @@ - (void)setComment:(NSString *)comment
-
 #pragma mark - Zip file detection
 
-
```
(Nitpick:) Please remove this whitespace change

abbeycode commented on Source/UZKArchive.m:
```diff
@@ -1973,6 +1999,11 @@ - (BOOL)performWriteAction:(int(^)(uLong *crc, NSError * __autoreleasing*innerEr
         zip_fileinfo zi = [UZKArchive zipFileInfoForDate:fileDate];
         
+        if (posixPermissions > 0) {
+            
```
(Nitpick:) Please remove this blank line

abbeycode commented on Source/UZKArchive.m:
```diff
@@ -1973,6 +1999,11 @@ - (BOOL)performWriteAction:(int(^)(uLong *crc, NSError * __autoreleasing*innerEr
         
+        if (posixPermissions > 0) {
+            
+            zi.external_fa = (posixPermissions + 32768) << 16;
```
Can you please define a constant for `32768` on the previous line to self-document its purpose? Also, just so I know, why do we need to add this constant and then bit shift?

MartinLau7 replied:
Since the permission value accessed by NSFileManager is octal, some offset conversion is performed.

abbeycode commented on Source/UZKFileInfo.h:
```diff
@@ -78,6 +78,9 @@ typedef NS_ENUM(NSInteger, UZKCompressionMethod) {
 
-
+/**
+ @brief posixPermissions of the file.
```
Please replace with a brief explanation of how to read the permissions (which NSNumber property to read, whether it's a bit mask, etc.

abbeycode commented on Source/UZKArchive.m:
```diff
@@ -2652,7 +2683,7 @@ + (zip_fileinfo)zipFileInfoForDate:(NSDate *)fileDate
     zi.tmz_date.tm_year = (uInt)date.year;
     zi.internal_fa = 0;
-    zi.external_fa = 0;
+    zi.external_fa = 2175008768;                // default value: 644
```
I'm not sure we should be defining a nonzero default here. What's the reason for it? Also, why use the decimal form instead of an octal like `0644U`? Then you wouldn't need the comment.

MartinLau7 replied:
Since the original default value generated by the zip package causes the unzipped file permissions to be raised to 755, this will cause a permission attack. So I modified the default value to lower the expectation（Sorry, my English is very bad.）

abbeycode replied:
I think 755 is fine as a default - universal permissions for owner, and read/exec for everyone else. I'm not sure what you mean by "attack", but all it means is that if the developer using UnzipKit doesn't specify permissions, the permissions will be pretty wide open on the extracted file. If someone is working on security-critical archives, it's going to be up to the developer to specify UnzipKit permissions. (Also, no need to apologize for your English - it's much better than my Chinese :-)

abbeycode replied:
A [comment I made below|https://github.com/abbeycode/UnzipKit/pull/84/files/7db8290a1129b75be39afddb055c1ae0303d2afa#r290028686} pertains to this thread.

MartinLau7 replied:
Under the default 755 permissions, some malicious programs will be able to obtain high-privilege access to the zip and malicious attacks. I think 644 is enough for normal programs to execute.
Follow-up here can do if else to try

MartinLau7 commented:
First of all, I am sorry that I have not made the correct modifications in the submitted code, I will improve these errors in the future.

abbeycode commented:
Hey Martin, no big deal, this is part of the process for a pull request. Let's work on this together. The PR isn't final - you can push more changes to your branch, and it'll update.

abbeycode commented on Source/UZKArchive.m:
```diff
@@ -1973,6 +1999,11 @@ - (BOOL)performWriteAction:(int(^)(uLong *crc, NSError * __autoreleasing*innerEr
+            
+            zi.external_fa = (posixPermissions + 32768) << 16;
+        }
+        
```
Perhaps it would make sense to put an else clause here, and assign `zi.external_fa` based on the file's actual permissions if nothing is specified. But if we do that, it changes behavior, and I would wait until after v1.9 for that. I would bump the version number to v2.0 so people would know to check. What do you think?

MartinLau7 replied:
I think it's okay, keep the actual permissions of the file without specifying content.

MartinLau7 commented on Source/UZKArchive.h:
```diff
@@ -618,6 +618,33 @@ compressionMethod:(UZKCompressionMethod)method
+compressionMethod:(UZKCompressionMethod)method
+         password:(nullable NSString *)password
+        overwrite:(BOOL)overwrite
+            error:(NSError **)error;
```
Hi abbeycode, i Add this method without changing the original code base until the code passes. Is this feasible?

MartinLau7 commented on Source/UZKArchive.m:
```diff
@@ -582,7 +582,7 @@ - (BOOL)extractFilesTo:(NSString *)destinationDirectory
                     BOOL createSuccess = [fm createFileAtPath:path
                                                      contents:nil
-                                                   attributes:nil];
+                                                   attributes:@{NSFilePosixPermissions:info.posixPermissions}];
```
Fix the actual permissions of the file not properly restored when unzipping

MartinLau7 commented:
@abbeycode There is currently a change in the package, I seem to be destroying the version of Xcode, it has been upgraded to Xcode 10.2

abbeycode commented:
Please let me know when you're done with your changes and are ready for me to review again. Also, make sure you review your changes in the "Files changed" tab when you're done, so they only contain changes you intended to make (not making whitespace changes to code you otherwise didn't touch).

Also, if you're having trouble understanding any of my comments, please let me know, and I can try to rephrase, or go into more detail.

abbeycode commented:
I just saw your Xcode 10.2 comment. The version of Xcode you use shouldn't matter, so long as you don't commit any upgrade checks or use APIs only available in newer versions of the frameworks (which Xcode shouldn't let you do unless you change the deployment target, which you shouldn't change either :)

abbeycode commented:
Did you see the list of requests I made in the [initial review comment](https://github.com/abbeycode/UnzipKit/pull/84#pullrequestreview-244270835)?

MartinLau7 commented:
> Did you see the list of requests I made in the [initial review comment](https://github.com/abbeycode/UnzipKit/pull/84#pullrequestreview-244270835)?

Yes, I see this problem now, because I have an exception in the code version of the workbench that I can't resubmit, I need to close the Pull Request later, until I rearrange and create a new Pull Request. What do you think?

abbeycode commented:
You don't need to close this PR, and you don't need to use a different version of Xcode. Just back out the changes you made to the `project.pbxproj` and `UnzipKit.xcscheme` files that aren't directly necessary for your change. You can use Xcode to this, if you view the changes from the v1.9 branch. Once you've backed out those changes, you can push to this branch again, and the PR will update.

MartinLau7 commented:
> You don't need to close this PR, and you don't need to use a different version of Xcode. Just back out the changes you made to the `project.pbxproj` and `UnzipKit.xcscheme` files that aren't directly necessary for your change. You can use Xcode to this, if you view the changes from the v1.9 branch. Once you've backed out those changes, you can push to this branch again, and the PR will update.

@abbeycode Since Xcode breaks git, it is not considered to turn off PR before. Now I have resubmitted and passed the Travis CI build, I will continue to modify if there are new issues.

abbeycode commented on Tests/TestKeppPrmissions.m:
```diff
@@ -0,0 +1,69 @@
+}
+
+- (void)testWriteFile {
+    
```
Since we're getting rid of `writeFile`, please replace this with a couple other tests that call `writeData`: One that specifies permissions, and one that doesn't. Then you can extract the archived file two ways in each test to verify the permissions: use `UZKArchive`, and also extract using the `zip` shell command, then verify the permissions of the extracted files using `NSFileManager`.

`-[UZKArchiveTestCase extractArchive:password]` is already written, and will perform the shell extraction. I can help if you have any trouble.

MartinLau7 replied:
Next, I may be unable to participate in the revision due to work problems. I need your help :(

abbeycode replied:
I can make a branch for this PR off of `v1.9` and put these test cases into that branch, if you'd like. You can then target that branch instead of `v1.9` for the PR, or at least rebase your changes off of that branch, so my changes become part of the PR. Your choice (either way is fine with me).

MartinLau7 replied:
I agree with you, this will help me to continue to change.

abbeycode commented on Tests/UZKArchiveTestCase.m:
```diff
@@ -61,7 +61,8 @@ - (void)setUp {
                            @"NotAZip-PK-ContentsUnknown",
-                           @"Modified CRC Archive.zip"];
+                           @"Modified CRC Archive.zip",
+                           @"Test Permissions Archive.zip",];
```
It might be better to generate the zip file instead of committing it to the repo. I've started doing this whenever I can so the repo size doesn't grow too much. It's pretty easy if you use one of the `-[UZKArchiveTestCase archiveWithFiles:]` overloads. I verified just now that the system `zip` shell command preserves permissions (on Catalina, at least, but I doubt it's changed). So you can write out some test text files and use `NSFileManager` to modify the permissions before archiving them. As with my suggestion above, I can help with this. I'd suggest writing a helper method in your test case class to write the files with the permissions you want, perhaps even naming the files with the octal permissions you're archiving (e.g. `777.txt`, `722.txt`).

MartinLau7 replied:
Next, I may be unable to participate in the revision due to work problems. I need your help :(

abbeycode commented on UnzipKit.xcodeproj/project.pbxproj:
```diff
@@ -447,6 +450,7 @@
 			developmentRegion = English;
 			hasScannedForEncodings = 0;
 			knownRegions = (
+				English,
```
Please revert this line (you must have missed it when you backed out the rest of your project changes).

MartinLau7 replied:
This is because Xcode is automatically generated, I try to recover, but there is still a problem. Maybe you can help me

abbeycode replied:
I'm sure Xcode is putting this here without your direct action. I'm just saying not to commit it to the repo. If you're using Xcode for your commits, you can uncheck this change. You can also revert it back outside of Xcode and commit.

MartinLau7 commented on Source/UZKArchive.h:
```diff
@@ -618,6 +618,32 @@ compressionMethod:(UZKCompressionMethod)method
+- (BOOL)writeData:(NSData *)data
+         filePath:(NSString *)filePath
+         fileDate:(nullable NSDate *)fileDate
+ posixPermissions:(unsigned long)posixPermissions
```
The type of posixPermissions usually depends on NSFileManager.

abbeycode replied:
Agreed. According to the documentation ([link](https://developer.apple.com/documentation/foundation/nsfileposixpermissions)), it's a short value, so let's use that.

MartinLau7 commented on Source/UZKArchive.m:
```diff
@@ -1164,9 +1167,52 @@ - (BOOL)writeData:(NSData *)data
+                  progress:nil
+                     error:error];
+#pragma clang diagnostic pop
+}
```
This was early reserved from the original version of Clone.

abbeycode replied:
Sure, that's fine. Please remove in your next commit.

MartinLau7 commented on Tests/TestKeppPrmissions.m:
```diff
@@ -0,0 +1,69 @@
+    NSError *extractError = nil;
+    BOOL success = [archive extractFilesTo:extractDirectory overwrite:NO error:&extractError];
+    XCTAssert(success, @"extract error %@", extractError);
+    // $ ls -l "extractURL.path" look up prmissions
```
The value recorded by NSFileManager has not been converted, so I recommend using ls -l to be clearer.

abbeycode replied:
I see your point, but I think we'd be better off writing our own method to use in the test cases to take the `NSFileManager` permissions value and format it as octal, for comparison purposes. One big reason is that we can't run shell commands (like `ls`) in iOS unit tests, and I prefer not making the test cases Mac-only if I can help it.

abbeycode commented:
I've created a new branch, `pr84`, to merge your PR into, for me to then work on it further. Can you please rebase your changes on that branch? I merged a different PR just now, that made two single-line fixes. After your rebase I'll merge into that branch and create a new PR for you to review once I've implemented the changes we discussed.

MartinLau7 commented:
> I've created a new branch, `pr84`, to merge your PR into, for me to then work on it further. Can you please rebase your changes on that branch? I merged a different PR just now, that made two single-line fixes. After your rebase I'll merge into that branch and create a new PR for you to review once I've implemented the changes we discussed. You can keep the "v1.9" branch name in your repo. That shouldn't matter.

I have rebase to "RP84", but there are some merge errors, I think I need to modify it in my free time, I will add some new code to optimize the "UZKFileInfo - isDirectory" implementation, then I will commit again.

abbeycode commented:
What are you changing in the `isDirectory` implementation?

Also, the changes I'm asking you to rebase into your branch are super simple: just a single commit affecting two lines ([link here](https://github.com/abbeycode/UnzipKit/commit/8c6f7403ef676a452fd6b96c027d6a34c8dc3c30)). Just resolve the rebase conflict using your version of `UZKArchive.m`, and before you commit make that same change, replacing `UZKCompressionMethodDefault` with `method` on those two lines.

MartinLau7 commented:
I have rebased to "PR84" and modified the previous error.
In the new commit I modified the isDirectory code to make it better for type judgment, and I added isSymLink and isResourceFork.
isSymLink can determine if the file is a symbolic link.
isResourceFork can help us decide whether to decompress the "__MACOSX" content.

MartinLau7 commented:
The new commit seems to have changed something in Xcode, and I have avoided submitting Xcode files as much as possible. This makes me very annoyed, I hope I can get your help and restore it to the original version.

abbeycode commented:
Martin, I would like to merge this PR into the `pr84` branch I created and make the remaining changes I'd like to see there myself. What would you like to get back to the original version, and what do you mean by the original version? There are ways in Git and in Xcode to checkout an older revision of a file over your working copy (or to compare its contents to the current revision), and then you can use Xcode to select which changes you want to commit (if any).

I'm going to squash and merge your PR and will start making my own commits on top. I'll open up a new PR and we can continue the discussion there.

abbeycode commented:
PR #87 was merged into the v1.9 branch, including these changes
